### ISSUE TITLE
Support checkpoint to prevent training from collapse

### ISSUE THREAD

**Issue by hijkzzz** [OpenLLMAI/OpenLLaMA2#65]

**Comment by hijkzzz** [comment]
@hijkzzz

**Comment by catqaq** [comment]
add basic ckpt function: https://github.com/OpenLLMAI/OpenRLHF/commit/f53571de43a4524644e75c9c472bbc69ac7b72c2

**Comment by karthik-nexusflow** [comment]
Hi team what are the next steps here ?, we can support this effort as we need this critically

**Comment by karthik-nexusflow** [comment]
from my understanding we need to support checkpointing and loading the actor and critic model

**Comment by hijkzzz** [comment]
@karthik-nexusflow 
We need to support the following features:
1. save and load actor and critic model weights optimizers schedulers, which can be done with DeepSpeed API
2. save and load the the progress of data loader, we may need to rewrite a new distributed sampler.
3. ssave and load seed

The second point can be tricky

**Comment by karthik19967829** [comment]
great for 2 , if we save all the dataset indices that were seen and skip that after we resume , could be a naive initial approach 

also we can initial support only for using 1 prompt dataset and not multiple datasets maybe

**Comment by hijkzzz** [comment]
> great for 2 , if we save all the dataset indices that were seen and skip that after we resume , could be a naive initial approach
> 
> also we can initial support only for using 1 prompt dataset and not multiple datasets maybe

Because the dataset will be shuffled before the training, we can skip the trained samples in the `for xxx in  dataloader` and also save/load the epoch.

**Comment by thirteenflt** [comment]
what's the status of this thread?
BTW it failed to save critic model ckpt in the current code with Ray. The error is:
File "/scratch/amlt_code/OpenRLHF/openrlhf/utils/deepspeed.py", line 409, in save_ckpt
    assert isinstance(model, deepspeed.DeepSpeedEngine)
AssertionError

Any idea?

**Comment by mickelliu** [comment]
> what's the status of this thread? BTW it failed to save critic model ckpt in the current code with Ray. The error is: File "/scratch/amlt_code/OpenRLHF/openrlhf/utils/deepspeed.py", line 409, in save_ckpt assert isinstance(model, deepspeed.DeepSpeedEngine) AssertionError
> 
> Any idea?

this is because the actor model is a local worker and the critic is a remote worker. You need to implement a function in the critic model class and call the remote method, for example, `critic_model.save_ckpt.remote()` ...

But that's a quick remedy, still need to wait for the official implementation. Personally, I don't find this feature very urgently required since training with OpenRLHF seems very stable, but rather I like to have intermediate models which I have implemented in my own fork.

**Comment by yangqiancheng-yuan** [comment]
> > great for 2 , if we save all the dataset indices that were seen and skip that after we resume , could be a naive initial approach
> > also we can initial support only for using 1 prompt dataset and not multiple datasets maybe
> 
> Because the dataset will be shuffled before the training, we can skip the trained samples in the `for xxx in dataloader` and also save/load the epoch.

Is this feature already supported? Does the resume training script now default to skipping the data that has already been trained?
